File: src/main/java/exnihilocreatio/registries/types/Meltable.java
package exnihilocreatio.registries.types;

import exnihilocreatio.util.BlockInfo;
import lombok.AllArgsConstructor;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.Setter;

import javax.annotation.Nonnull;

@AllArgsConstructor
@EqualsAndHashCode(exclude = {"textureOverride"})
public class Meltable {

    public static final Meltable EMPTY = new Meltable("", 0);

    @Getter
    private String fluid;

    @Getter
    @Setter
    private int amount;

    @Getter
    @Nonnull
    private BlockInfo textureOverride;

    public Meltable(String fluid, int amount) {
        this.fluid = fluid;
        this.amount = amount;
        this.textureOverride = BlockInfo.EMPTY;
    }

    public Meltable copy() {
        return new Meltable(fluid, amount, textureOverride);
    }

    public Meltable setTextureOverride(BlockInfo textureOverride) {
        this.textureOverride = textureOverride;
        return this;
    }
}

File: src/main/java/exnihilocreatio/registries/types/Meltable.kt
package exnihilocreatio.registries.types

import exnihilocreatio.util.BlockInfo

data class Meltable @JvmOverloads constructor(
        val fluid: String,
        var amount: Int,
        var textureOverride: BlockInfo = BlockInfo.EMPTY
) {
    fun copy(): Meltable {
        return Meltable(fluid, amount, textureOverride)
    }

    fun setTextureOverrideChain(textureOverride: BlockInfo): Meltable {
        this.textureOverride = textureOverride
        return this
    }

    companion object {
        val EMPTY = Meltable("", 0)
    }
}
